// @fluentui/react/lib-amd/components/GroupedList/GroupShowAll.styles.js
define(["require", "exports", "../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = void 0;
    var GlobalClassNames = {
        root: 'ms-GroupShowAll',
        link: 'ms-Link',
    };
    var getStyles = function (props) {
        var _a;
        var theme = props.theme;
        var fonts = theme.fonts;
        var classNames = (0, Styling_1.getGlobalClassNames)(GlobalClassNames, theme);
        return {
            root: [
                classNames.root,
                {
                    position: 'relative',
                    padding: '10px 84px',
                    cursor: 'pointer',
                    selectors: (_a = {},
                        _a[".".concat(classNames.link)] = {
                            fontSize: fonts.small.fontSize,
                        },
                        _a),
                },
            ],
        };
    };
    exports.getStyles = getStyles;
});
