// @fluentui/react/lib-amd/components/ActivityItem/ActivityItem.classNames.js
define(["require", "exports", "../../Styling", "../../Utilities"], function (require, exports, Styling_1, Utilities_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getClassNames = void 0;
    exports.getClassNames = (0, Utilities_1.memoizeFunction)(function (styles, className, activityPersonas, isCompact) {
        return {
            root: (0, Styling_1.mergeStyles)('ms-ActivityItem', className, styles.root, isCompact && styles.isCompactRoot),
            pulsingBeacon: (0, Styling_1.mergeStyles)('ms-ActivityItem-pulsingBeacon', styles.pulsingBeacon),
            personaContainer: (0, Styling_1.mergeStyles)('ms-ActivityItem-personaContainer', styles.personaContainer, isCompact && styles.isCompactPersonaContainer),
            activityPersona: (0, Styling_1.mergeStyles)('ms-ActivityItem-activityPersona', styles.activityPersona, isCompact && styles.isCompactPersona, !isCompact && activityPersonas && activityPersonas.length === 2 && styles.doublePersona),
            activityTypeIcon: (0, Styling_1.mergeStyles)('ms-ActivityItem-activityTypeIcon', styles.activityTypeIcon, isCompact && styles.isCompactIcon),
            activityContent: (0, Styling_1.mergeStyles)('ms-ActivityItem-activityContent', styles.activityContent, isCompact && styles.isCompactContent),
            activityText: (0, Styling_1.mergeStyles)('ms-ActivityItem-activityText', styles.activityText),
            commentText: (0, Styling_1.mergeStyles)('ms-ActivityItem-commentText', styles.commentText),
            timeStamp: (0, Styling_1.mergeStyles)('ms-ActivityItem-timeStamp', styles.timeStamp, isCompact && styles.isCompactTimeStamp),
        };
    });
});
